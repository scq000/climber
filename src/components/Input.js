import React from 'react';
import {View, Animated, TextInput, StyleSheet} from 'react-native';

const DURATION_TIME = 150;
const FOCUS_VALUE = 1;
const BLUR_VALUE = 0;

const Input = ({label, keyboardType, onChangeText, value = '', ...rest}) => {
  const [isActive, setIsActive] = React.useState(false);

  const positionAnim = React.useRef(
    new Animated.Value(value ? FOCUS_VALUE : BLUR_VALUE),
  ).current;

  const _onFocus = () => {
    if (!isActive) {
      setIsActive(true);
      Animated.timing(positionAnim, {
        toValue: FOCUS_VALUE,
        duration: DURATION_TIME,
        useNativeDriver: false,
      }).start();
    }
  };

  const _onBlur = () => {
    if (isActive && !value) {
      setIsActive(false);
      Animated.timing(positionAnim, {
        toValue: BLUR_VALUE,
        duration: DURATION_TIME,
        useNativeDriver: false,
      }).start();
    }
  };

  const _getAnimatedTitleStyles = () => {
    const titleActiveSize = 12;
    const titleInActiveSize = 16;

    return {
      top: positionAnim.interpolate({
        inputRange: [0, 0.75],
        outputRange: [14, 0],
      }),
      fontSize: isActive ? titleActiveSize : titleInActiveSize,
    };
  };

  return (
    <View style={styles.container}>
      <Animated.Text style={[styles.title, _getAnimatedTitleStyles()]}>
        {label}
      </Animated.Text>
      <TextInput
        value={value}
        underlineColorAndroid="transparent"
        onFocus={_onFocus}
        onBlur={_onBlur}
        onChangeText={onChangeText}
        keyboardType={keyboardType}
        {...rest}
        style={[styles.textInput, rest.style]}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    width: '100%',
    borderBottomColor: '#7b7b7b',
    borderBottomWidth: 0.5,
    borderStyle: 'solid',
    marginVertical: 5,
  },
  title: {
    position: 'absolute',
    left: 4,
    color: '#7b7b7b',
  },
  textInput: {
    fontFamily: 'Helvetica',
    color: '#fff',
    paddingBottom: 0,
  },
});

export default Input;
