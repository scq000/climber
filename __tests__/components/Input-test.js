import 'react-native';
import React from 'react';
import {create} from 'react-test-renderer';
import Input from '../../src/components/Input';

describe('Input', () => {
  it('should render correctly', function () {
    const snap = create(<Input />).toJSON();
    expect(snap).toMatchSnapshot();
  });

  it('should render with label correctly', function () {
    const snap = create(<Input label="test-label" />).toJSON();
    expect(snap).toMatchSnapshot();
  });
});