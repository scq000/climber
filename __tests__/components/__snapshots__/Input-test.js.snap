// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render correctly 1`] = `
<View
  style={
    Object {
      "borderBottomColor": "#7b7b7b",
      "borderBottomWidth": 0.5,
      "borderStyle": "solid",
      "marginVertical": 5,
      "width": "100%",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#7b7b7b",
        "fontSize": 16,
        "left": 4,
        "position": "absolute",
        "top": 14,
      }
    }
  />
  <TextInput
    allowFontScaling={true}
    onBlur={[Function]}
    onFocus={[Function]}
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "color": "#fff",
          "fontFamily": "Helvetica",
          "paddingBottom": 0,
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;

exports[`Input should render with label correctly 1`] = `
<View
  style={
    Object {
      "borderBottomColor": "#7b7b7b",
      "borderBottomWidth": 0.5,
      "borderStyle": "solid",
      "marginVertical": 5,
      "width": "100%",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#7b7b7b",
        "fontSize": 16,
        "left": 4,
        "position": "absolute",
        "top": 14,
      }
    }
  >
    test-label
  </Text>
  <TextInput
    allowFontScaling={true}
    onBlur={[Function]}
    onFocus={[Function]}
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "color": "#fff",
          "fontFamily": "Helvetica",
          "paddingBottom": 0,
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;
